File: Assets/Scripts/C#/Manager/EmailManager/EmailInfo.cs
using System;
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

public enum EmailStatus
{
    Finish = 0,
    haveCheck = 1,
    haveItem = 2,
}

public class EmailInfo
{
    //邮件ID
    public int emailId;
    //发放条件
    public int conditionId;
    //邮件标题
    public string emailTitle;
    //邮件正文
    public string emailBody;
    //发件人
    public string emailSender;
    //发件时间
    public DateTime emailSendTime;
    //有效时间
    public long lostTimer;
    //邮件奖励道具ID列表
    public List<int> itemIdLis = new List<int>();
    //邮件状态
    public EmailStatus emailStatus;
    //是否有道具
    public bool hasItem = false;

    public EmailInfo() { }

    public EmailInfo(int emailId, 
                     int conditionId, 
                     string emailTitle, 
                     string emailSender, 
                     string emailBody, 
                     DateTime emailSendTime, 
                     string lostTimer, 
                     string itemIdLis, EmailStatus status)
    {
        this.emailId = emailId;
        this.conditionId = conditionId;
        this.emailTitle = emailTitle;
        this.emailSender = emailSender;
        this.emailBody = emailBody;
        this.emailSendTime = emailSendTime;
        this.lostTimer = long.Parse(lostTimer);

        string[] itemIds = itemIdLis.Split(new char[] { '|' });
        for (int i = 0; i < itemIds.Length; i++)
        {
            if (itemIds[i]=="")break;
            this.itemIdLis.Add(int.Parse(itemIds[i]));
        }
        emailStatus = status;
        if(itemIdLis.Length > 0)
        {
            hasItem = true; 
        }
    }

    /// <summary>
    /// 变更EmailInfo的 Status
    /// </summary>
    public bool ChangeStatus()
    {
        this.emailStatus = EmailStatus.Finish;
        return isOverTime();
    }

    public bool isOverTime()
    {
        DateTime clientTime = DateTime.Now;
        TimeSpan ts = clientTime - emailSendTime;
        long lastTime = (long)(lostTimer - ts.TotalSeconds);
        if (lastTime <= 0)
        {
            Debug.LogError("邮件已过期");
            return false;
        }
        return true;
    }
}

File: Assets/Scripts/LuaScripts/UI/Info/EmailInfo.lua
EmailStatus = {
    Finish = 0,
    haveCheck = 1,
    haveItem = 2
}

local EmailInfo = BaseClass.Create()

function EmailInfo:__init(email)
    self:init(  email.emailId,email.conditionId,email.emailTitle,email.emailSender,
                email.emailBody,email.emailSendTime,email.lostTimer,email.itemIdLis,email.emailStatus)
end


function EmailInfo:init( emailId, conditionId, emailTitle, emailSender, emailBody, emailSendTime, lostTimer, itemIdLis, status )
    self.emailId = emailId
    self.conditionId = conditionId
    self.emailTitle = emailTitle
    self.emailSender = emailSender
    self.emailBody = emailBody
    self.emailSendTime = self:TimeTranslate(emailSendTime)
    self.lostTimer = lostTimer
    self.emailStatus = status

    self.itemIdLis = {}
    -- for itemId in itemIdLis:gmatch("[^|]+") do
    --     table.insert(itemIdLis, itemId)
    -- end
    for i = 0, itemIdLis.Count - 1 do
        table.insert(self.itemIdLis, itemIdLis[i])
    end

    self.hasItem = #self.itemIdLis ~= 0 

end

function EmailInfo:TimeTranslate(emailSendTime)

    local timeTable = {}
    
    timeTable["year"] = emailSendTime.Year;
    timeTable["month"] = emailSendTime.Month;
    timeTable["day"] = emailSendTime.Day;
    timeTable["hour"] = emailSendTime.Hour;
    timeTable["min"] = emailSendTime.Minute;
    timeTable["sec"] = emailSendTime.Second;

    local translatedTime = os.time(timeTable)

    timeTable = nil

    return translatedTime
end

function EmailInfo:ChangeStatus()
    self.emailStatus = EmailStatus.Finish
    return self:isOverTime()
end

function EmailInfo:isOverTime()
    local currentTime = os.time()
    local timeSpan = currentTime - self.emailSendTime
    local lastTime = self.lostTimer - timeSpan

    if lastTime <= 0 then
        print(self.emailId.." ["..self.emailTitle.."] : ".."邮件已过期")
        return false
    end

    return true
end

return EmailInfo
